File: metrics.py
from typing import List
import numpy as np
import pandas as pd


def compute_sharpe_ratio(returns: List[float], risk_free_rate: float = 0.0) -> float:

    """
    Compute the annualized Sharpe ratio of a portfolio.

    Args:
        returns (List): A list of returns for the portfolio.
        risk_free_rate (float): The risk-free rate.

    Returns:
        float: The annualized Sharpe ratio of the portfolio.
    """

    if len(returns) == 0:
        return -1

    # Calculate the mean and standard deviation of the returns
    mean_return = np.mean(returns)
    std_return = np.std(returns)

    if std_return == 0:
        return 0

    # Calculate the annualized Sharpe ratio
    sharpe_ratio = (mean_return - risk_free_rate) / std_return # * np.sqrt(len(returns))

    # if sharpe_ratio > 6:
    #     print(f"returns: {returns} \n mean_return: {mean_return} | std_return: {std_return} | sharpe_ratio: {sharpe_ratio}")

    return sharpe_ratio


def compute_cagr_from_returns(returns: List[float], from_time: pd.Timestamp, to_time: pd.Timestamp, initial_value: float = 1.0) -> float:

    years = (to_time - from_time).days / 365.25
    if years <= 0:
        return 0

    # final_value = np.nanprod([1 + r for r in returns])
    final_value = initial_value
    for r in returns:
        if not np.isnan(r):
            final_value *= (1 + r / 100)
    if final_value <= 0:
        return -1

    cagr = (final_value / initial_value) ** (1 / years) - 1

    # print(f"returns: {returns}")
    # print(f"years: {years} | initial_value:  {round(initial_value, 2)} | final_value:  {round(final_value, 2)}")

    return cagr

File: backtestingCpp/Metrics.cpp
#include <cmath>
#include <numeric>
#include <limits>

#include "Metrics.h"

// Sharpe Ratio = (Mean Return - Risk-Free Rate) / Std Dev of Returns
double compute_sharpe_ratio(const std::vector<double>& returns, double risk_free_rate)
{
    if (returns.empty()) return 0.0;

    double mean = std::accumulate(returns.begin(), returns.end(), 0.0) / returns.size();

    double variance = 0.0;
    for (double r : returns)
        variance += (r - mean) * (r - mean);
    variance /= returns.size();

    double std_dev = std::sqrt(variance);
    if (std_dev == 0.0) return 0.0;

    return (mean - risk_free_rate) / std_dev;
}

// CAGR using total % return and years elapsed
double compute_cagr_from_percent(double total_return_percent, long long from_time, long long to_time)
{
    double years = compute_years_from_ms(from_time, to_time);
    if (years <= 0.0)
        return 0.0;

    double multiplier = 1.0 + total_return_percent / 100.0;

    if (multiplier <= 0.0)
        return -1.0; // Represents complete loss or worse

    return std::pow(multiplier, 1.0 / years) - 1.0;
}

double compute_cagr_from_returns(const std::vector<double>& returns, long long from_time, long long to_time, double initial_value)
{
    double years = compute_years_from_ms(from_time, to_time);
    if (years <= 0.0)
        return 0.0;
    double final_value = initial_value;
    for (double r : returns) {
        final_value *= (1 + r / 100);
    }
    if (final_value <= 0) {
        return -1.0;
    }
    
    double cagr = std::pow((final_value / initial_value), (1.0 / years)) - 1.0;
    return cagr;
}

// Convert millisecond timestamps to fractional years
double compute_years_from_ms(long long from_time, long long to_time)
{
    return static_cast<double>(to_time - from_time) / 31536000000.0;  // 365 days
}
